server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```
@@ -46,42 +75,112 @@ public void beforeRefresh() throws IOException {
+                            List<String> segmentInfosFiles = refreshedLocalFiles.stream()
+                                .filter(file -> file.startsWith(IndexFileNames.SEGMENTS))
+                                .collect(Collectors.toList());
+                            Optional<String> latestSegmentInfos = segmentInfosFiles.stream()
+                                .max(Comparator.comparingLong(IndexFileNames::parseGeneration));
+
+                            if (latestSegmentInfos.isPresent()) {
+                                refreshedLocalFiles.addAll(SegmentInfos.readCommit(storeDirectory, latestSegmentInfos.get()).files(true));
+                                segmentInfosFiles.stream()
+                                    .filter(file -> !file.equals(latestSegmentInfos.get()))
+                                    .forEach(refreshedLocalFiles::remove);
+
+                                boolean uploadStatus = uploadNewSegments(refreshedLocalFiles);
+                                if (uploadStatus) {
+                                    remoteDirectory.uploadMetadata(
+                                        refreshedLocalFiles,
+                                        storeDirectory,
+                                        indexShard.getOperationPrimaryTerm(),
+                                        segmentInfos.getGeneration()
+                                    );
+                                    localSegmentChecksumMap.keySet()
+                                        .stream()
+                                        .filter(file -> !refreshedLocalFiles.contains(file))
+                                        .collect(Collectors.toSet())
+                                        .forEach(localSegmentChecksumMap::remove);
+                                }
+                            }
+                        } catch (EngineException e) {
```

Since we swallow all exceptions, what happens if shard was promoted to primary but segments were not refreshed due to some failures? Should we go back to replica mode?

---

With the current implementation, we retry segment upload in the next refresh.

If the upload fails, we don't want to demote primary as the failure could be due to issues with remote segment store itself.